File: pinyougou-web/pinyougou-user-web/src/main/java/com/pinyougou/user/controller/UserController.java
package com.pinyougou.user.controller;

import com.alibaba.dubbo.config.annotation.Reference;
import com.pinyougou.pojo.*;
import com.pinyougou.service.UserService;
import org.apache.commons.codec.digest.DigestUtils;
import com.pinyougou.service.*;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletRequest;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import java.text.SimpleDateFormat;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 用户控制器
 *
 * @author lee.siu.wah
 * @version 1.0
 * <p>File Created at 2019-04-15<p>
 */
@RestController
@RequestMapping("/user")
public class UserController {

    @Reference(timeout = 10000)
    private UserService userService;

    @Reference(timeout = 10000)
    private ProvincesService provincesService;
    @Reference(timeout = 10000)
    private CitiesService citiesService;
    @Reference(timeout = 10000)
    private AreasService areasService;

    /** 用户注册 */
    @PostMapping("/save")
    public boolean save(@RequestBody User user, String code){
        try{
            // 检验验证码是否正确
            boolean flag = userService.checkSmsCode(user.getPhone(), code);
            if (flag) {
                userService.save(user);
            }
            return flag;
        }catch (Exception ex){
            ex.printStackTrace();
        }
        return false;
    }


    /** 发送短信验证码 */
    @GetMapping("/sendSmsCode")
    public boolean sendSmsCode(String phone){
        try{
            return userService.sendSmsCode(phone);
        }catch (Exception ex){
            ex.printStackTrace();
        }
        return false;
    }

    @GetMapping("/findUserByUsername")
    public Map<String,Object> findUserByUsername(String username){
        Map<String,Object > map = new HashMap<>();
        try{
            User user = userService.findUserByUsername(username);
            map.put("user",user);
            map.put("birthdayString",new SimpleDateFormat("yyyy-MM-dd").format(user.getBirthday() == null ? new Date():user.getBirthday()));
            return map;
        }catch (Exception ex){
            ex.printStackTrace();
        }
        return null;
    }
    //修改的方法
    @PostMapping("/updateAddress")
    public Boolean updateAddress(@RequestBody Address address){
        try {
            userService.updateAddress(address);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }
    //保存的方法

    @PostMapping("/saveAddress")
    public Boolean saveAddress(@RequestBody Address address,HttpServletRequest request){
        try {
            address.setUserId(request.getRemoteUser());
            userService.saveAddress(address);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }
    @GetMapping("/findProvincesList")
    public List<Provinces> findProvincesList(){
        try{
            return provincesService.findAll();
        }catch (Exception ex){
            ex.printStackTrace();
        }
        return null;
    }
    @GetMapping("/deleteAddress")
    public Boolean deleteAddress(Long id){
        try {
            userService.deleteAddress(id);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }
    @GetMapping("/findCitiesList")
    public List<Cities> findCitiesList(String provinceId){
        try{
            System.out.println(provinceId);
            return citiesService.findByProvinceId(provinceId);
        }catch (Exception ex){
            ex.printStackTrace();
        }
        return null;
    }

    @GetMapping("/findAreasList")
    public List<Areas> findAreasList(String cityId){
        try{
            return areasService.findByCityId(cityId);
        }catch (Exception ex){
            ex.printStackTrace();
        }
        return null;
    }

    @PostMapping("/saveUser")
    public Boolean saveUser(@RequestBody User entity){
        try{
            userService.saveUser(entity);
            return true;
        }catch (Exception ex){
            ex.printStackTrace();
        }
        return false;
    }

    //修改密码
    @PostMapping("/codePassword")
    public Boolean codePassword(@RequestBody Map userMap,HttpServletRequest request){
        try {
            String oldPassword = (String) userMap.get("oldPassword");
            String username = request.getRemoteUser();
            User user = userService.selectUser(username);
            String password = user.getPassword();
            if (DigestUtils.md5Hex(oldPassword).equals(password)){
                userService.updatePassword((String)userMap.get("newPassword"),username);
                return true;
            }else {
                return false;
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @GetMapping("/findAddressByParentId")
    public List<Provinces> findAllProvince(){
        try {
            return userService.findAllProvince();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
    @GetMapping("/findCitiesByProvinceId")
    public List<Cities> findCitiesByProvinceId(String provinceId){
        try {
            return userService.findAllCitiesByProvinceId(provinceId);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    @GetMapping("/findAreasByCityId")
    public List<Areas> findAreasByCityId(String cityId){
        try {
            return userService.findAreasByCityId(cityId);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }
    @GetMapping("/findUserPhoneByUserId")
    public String findUserPhoneByUserId(HttpServletRequest request){
        try {
            return userService.findUserPhoneByUserId(request.getRemoteUser());
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    @GetMapping("/judgeCode")
    public Boolean judgeCode(String code,HttpServletRequest request){
        try {
            String ve = (String) request.getSession().getAttribute(VerifyController.VERIFY_CODE);
            if (ve.equals(code)){
                return true;
            }

        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    @GetMapping("/sendCode")
    public Boolean sendCode(String phone){
        try {
            userService.sendCode(phone);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    @GetMapping("/checkSmsCode")
    public Boolean checkSmsCode(String phone,String code,HttpServletRequest request){
        try {
           if(userService.checkSmsCode(phone, code)){
               userService.saveVerifyPhone(request.getRemoteUser(),"true");
                return true;
            };
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }


    @GetMapping("/updatePhone")
    public Map<String, Object> updatePhone(String phone,String code,HttpServletRequest request){
        Map<String,Object> map = new HashMap<>();
        try {
            if (!userService.verifyPhone(request.getRemoteUser(), "true")){
                map.put("success",false);
                map.put("message","请先验证手机号");
                return map;
            }
            if(userService.checkSmsCode(phone, code)){
                User user = new User();
                user.setPhone(phone);
                user.setUsername(request.getRemoteUser());
                user.setUpdated(new Date());
                userService.updatePhone(user);
                map.put("success",true);
                return map;
            } else {
                map.put("success",false);
                map.put("message","短信验证码不正确");
                return map;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        map.put("success",false);
        map.put("message","网络出现异常");
        return map;
    }



    //修改的方法
    @PostMapping("/updateAddress")
    public Boolean updateAddress(@RequestBody Address address){
        try {
            userService.updateAddress(address);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    //保存的方法

    @PostMapping("/saveAddress")
    public Boolean saveAddress(@RequestBody Address address,HttpServletRequest request){
        try {
            address.setUserId(request.getRemoteUser());
            userService.saveAddress(address);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    @GetMapping("/deleteAddress")
    public Boolean deleteAddress(Long id){
        try {
            userService.deleteAddress(id);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }
}

File: pinyougou-web/pinyougou-user-web/src/main/webapp/js/controller/userController.js
/** 定义控制器层 */
app.controller('userController', function($scope, $timeout, baseService,$controller){
    $controller('indexController',{$scope:$scope});
    // 定义user对象
    $scope.user = {};
    /** 用户注册 */
    $scope.save = function () {

        // 判断两次密码是否一致
        if ($scope.okPassword && $scope.user.password == $scope.okPassword){
            // 发送异步请求
            baseService.sendPost("/user/save?code=" + $scope.code, $scope.user)
                .then(function(response){
                if (response.data){
                    // 清空表单数据
                    $scope.user = {};
                    $scope.okPassword = "";
                    $scope.code = "";
                }else{
                    alert("注册失败！");
                }
            });

        }else{
            alert("两次密码不一致！");
        }
    };




    // 发送短信验证码
    $scope.sendSmsCode = function () {

        // 判断手机号码
        if ($scope.user.phone && /^1[3|4|5|7|8]\d{9}$/.test($scope.user.phone)){
            // 发送异步请求
            baseService.sendGet("/user/sendSmsCode?phone=" + $scope.user.phone)
                .then(function(response){
                if (response.data){
                    // 调用倒计时方法
                    $scope.downcount(90);

                }else{
                    alert("发送失败！");
                }
            });
        }else {
            alert("手机号码格式不正确！")
        }
    };


    $scope.smsTip = "获取短信验证码";
    $scope.disabled = false;

    // 倒计时方法
    $scope.downcount = function (seconds) {

        seconds--;

        if (seconds >= 0){
            $scope.smsTip = seconds + "秒后，重新获取！";
            $scope.disabled = true;
            // 第一个参数：回调的函数
            // 第二个参数：间隔的时间毫秒数
            $timeout(function(){
                $scope.downcount(seconds);
            }, 1000);
        }else {
            $scope.smsTip = "获取短信验证码";
            $scope.disabled = false;
        }

    };


    //判断用户原密码是否正确
    $scope.changePassword=function () {
        if($scope.user.newPassword == $scope.rePassword && $scope.user.newPassword){
            if ($scope.user.oldPassword ){
                baseService.sendPost("/user/codePassword",$scope.user).then(function (response) {
                    if (response.data == true){
                        alert("修改成功");
                        $scope.user={};
                        $scope.rePassword="";
                    }else {
                        $("#oldPasswordWarn").html("原密码错误!");
                    }
                })
            }else {
                $("#oldPasswordWarn").html("密码不能为空!");
            }
        }else {
            $("#rePasswordWarn").html("修改的密码有误!");
            return;
        }
    };
    // 获取收件人地址列表
    $scope.findAddressByUser = function () {
        // 发送异步请求
        baseService.sendGet("/address/findAddressByUser").then(function(response){
            // 获取响应数据
            $scope.addressList = response.data;

            // 获取默认的收件地址
            $scope.address = $scope.addressList[0];
        });
    };

    //显示详细地址
    $scope.showAddress=function (item) {
        $scope.addressSave=item;
    };
    $scope.changeAlias=function (alias) {
        $scope.addressSave.alias=alias;
    };
    //根据父级id查询地区
    $scope.findAddressByParentId=function () {
        baseService.sendGet("/user/findAddressByParentId").then(function (response) {
            $scope.Provinces = response.data;
        })
    };

   
   $scope.$watch("addressSave.provinceId",function (newValue,oldValue) {
        if($scope.addressSave.provinceId){
            baseService.sendGet("/user/findCitiesByProvinceId?provinceId="+newValue).then(function (response) {
                $scope.cities = response.data;
            })
        }
    });

    $scope.$watch("addressSave.cityId",function (newValue,oldValue) {
        if($scope.addressSave.cityId){
            baseService.sendGet("/user/findAreasByCityId?cityId="+newValue).then(function (response) {
                $scope.areas = response.data;

            })
        }
    })
});